### ISSUE TITLE
start private chat

### ISSUE THREAD

**Issue by icubed-web** [socketio/socket.io#2177]
Hi ,

I would like to start a private chat between two users 

   user1 = logged on user
  user 2 = site visitor

When user1 logs on a socket for that user will be create .

When user2 browses user1 profile they can initiate a private chat .

I know this can be done with rooms but for the life of me I dont know how to get this done .

Any help will be greatly appreciated .

Thankd

**Comment by knusprigeswuerstchen** [comment]
You could just create a private Room and joining their sockets to that room.

Example:

    var users = [
       { id: "3535" },
       { id: "7373" }
    ]
    var room = users[0].id + users[1].id;
    io.on('connection', function(socket){
       socket.join(room);
    });

**Comment by icubed-web** [comment]
client.on('connection', function (socket) {
  
	sockets.push(socket);
    //join the server
    socket.on('join', function(name){
      people[socket.id] = name;
      console.log(people);
    });

    //disconnect from the server
    socket.on('disconnect', function(){
      delete people[socket.id];
      sockets.splice(sockets.indexOf(socket), 1);
    });

    //initiate private message
    socket.on('initiate private message',function(data){
    	var nae = data.name;
    	console.log(nae);
      var receiverSocketId = findUserByName(nae);
      if(receiverSocketId){
        var receiver = people[receiverSocketId];
        var room = getARoom(people[socket.id], receiver);

        //join the anonymous user
        socket.join(room);
        //join the registered user 
        sockets[receiverSocketId].join(room);


        //notify the client of this
        socket.in(room).emit('private room created', room);

        //
      }
    });

		    socket.on('send private message', function(room, message){
		      socket.in(room).emit('private chat message', message);
		    });
  });

//you could use e.g. underscore to achieve this (
function findUserByName(name){
  for(socketID in people){
    if(people[socketID].name === name){
      return test = socketID;

     

    }
  }
  // return false;
console.log('not there');
}

//generate private room name for two users
function getARoom(user1, user2){
  return 'privateRooom' + user1.name + "And" + user2.name;
}


Im getting stuck at : 

       //join the anonymous user
        socket.join(room);
        //join the registered user 
        sockets[receiverSocketId].join(room);

**Comment by knusprigeswuerstchen** [comment]
Please format your code ([Syntax Highlighting](https://guides.github.com/features/mastering-markdown/#GitHub-flavored-markdown)) like this:

````javascript
sockets.push(socket);
//join the server
socket.on('join', function(name){
  people[socket.id] = name;
  console.log(people);
});

//disconnect from the server
socket.on('disconnect', function(){
  delete people[socket.id];
  sockets.splice(sockets.indexOf(socket), 1);
});

//initiate private message
socket.on('initiate private message',function(data){
    var nae = data.name;
    console.log(nae);
  var receiverSocketId = findUserByName(nae);
  if(receiverSocketId){
    var receiver = people[receiverSocketId];
    var room = getARoom(people[socket.id], receiver);

    //join the anonymous user
    socket.join(room);
    //join the registered user 
    sockets[receiverSocketId].join(room);


    //notify the client of this
    socket.in(room).emit('private room created', room);

    //
  }
});

        socket.on('send private message', function(room, message){
          socket.in(room).emit('private chat message', message);
        });
});

//you could use e.g. underscore to achieve this (
function findUserByName(name){
for(socketID in people){
if(people[socketID].name === name){
return test = socketID;

}
}
// return false;
console.log('not there');
}

//generate private room name for two users
function getARoom(user1, user2){
return 'privateRooom' + user1.name + "And" + user2.name;
}
````

I don't understand why you are stuck there. Could you paste the error from log?

**Comment by icubed-web** [comment]
Sorry , 

    at Socket.<anonymous> (C:\Users\..\s
erver.js:38:34)
    at Socket.emit (events.js:107:17)
    at Socket.onevent (C:\Users\..\node_
modules\socket.io\lib\socket.js:330:8)
    at Socket.onpacket (C:\Users\..\node
_modules\socket.io\lib\socket.js:290:12)
    at Client.ondecoded (C:\Users\..\nod
e_modules\socket.io\lib\client.js:193:14)
    at Decoder.Emitter.emit (C:\Users\..\
\node_modules\socket.io\node_modules\socket.io-parser\node_modules\component-emi
tter\index.js:134:20)
    at Decoder.add (C:\Users\..\node_mod
ules\socket.io\node_modules\socket.io-parser\index.js:247:12)
    at Client.ondata (C:\Users\..\node_m
odules\socket.io\lib\client.js:175:18)
    at Socket.emit (events.js:107:17)
    at Socket.onPacket (C:\Users\..\node
_modules\socket.io\node_modules\engine.io\lib\socket.js:99:14)

I do appreciate your help on this , as Im trying to get this going for the past 4 days .

Again Much appreciated .

**Comment by icubed-web** [comment]
And this is my client code 

``````````````

(function() {
		var getNode = function(s) {

			return document.querySelector(s);
		},

		messages = getNode('.chat-messages'),
		textarea = getNode('.chat textarea'),
	 	chatName = getNode('.chat-name'),
	 	join	 = getNode('#join');

	 		try	{

			var socket = io.connect('http://127.0.0.1:8080');

		}catch (e) {

		}

						socket.on('send private message', function(data){
					if (data.length) {
					// Loop through results
						for (var x = 0; x < data.length; x = x + 1) {
						var message = document.createElement('div');
						message.setAttribute('class', 'chat-message');
						message.textContent = data[x].name + ': ' + data[x].message;

	

						// Append
					
						messages.appendChild(message);
						messages.insertBefore(message, messages.firstChild);
		
					}
				}

			});

				join.addEventListener('click', function(event){
				
					name = chatName.value;

					socket.emit('join', {
						name: name
					});
	
			});



				textarea.addEventListener('keydown', function(event){
					var self = this ,
					name = chatName.value;
					
					socket.emit('initiate private message', {
						name: name,
						message: self.value

					});


			});


	})();

``````````

**Comment by knusprigeswuerstchen** [comment]
Thank you. I look over it and try to find the error.
Could you paste me the lines mentioned in the error log?

**Comment by icubed-web** [comment]
Sorry new to node.js ,

How would I get the errors log ,

Im using localStorage.debug = '*';

**Comment by knusprigeswuerstchen** [comment]
You could solve this problem like its done [here](http://socket.io/docs/rooms-and-namespaces/#default-room),
e.g.
````javascript
  //private message snippet
  socket.on('initiate private message', function(id, msg){
    //id is the receivers socket id
    socket.broadcast.to(id).emit('private chat created', msg);
  });
```

**Comment by icubed-web** [comment]
Im still receiving the same errors as before .

Maybe if I give you a run down of what imm trying to do it should help .

First when the document loads , I enter a username on browser 1 and click join - which has socket.on('join') .

Next in a separate window I enter another username and click join .

When I try entering any data in the textarea the error gets outputted .

Im sorry if this is confusing .

**Comment by paradite** [comment]
@icubed-web are you still working on this?

**Comment by 1pk73** [comment]
Hello I wanted to add this functionality to my site.Any help would be praising.

**Comment by chb7** [comment]
## __

**Comment by RagOfJoes** [comment]
I've just started developing my own chat application with its own private chat functionality. A solution that I came up with was creating a message schema that includes both user's username.  So then when a user chooses to chat with another user then I will have an endpoint attempting to search through my messages document and if none are found with said user's username then I will upsert a new model(create a new model) then use that model's generated ID for the room's ID.

I know this post is pretty old but for any lurkers on here that just want to get an idea for an approach to this problem there you go.
